File: lib/ptabler/software/src/ptabler/expression/basics.py
from cmath import log
import typing
import operator
import polars as pl

from ptabler.common import PType, toPolarsType

from .base import Expression

# Comparison Expressions

AnyExpression = Expression


class BinaryOperatorExpression(Expression):
    lhs: 'AnyExpression'
    rhs: 'AnyExpression'


class GtExpression(BinaryOperatorExpression, tag='gt'):
    def to_polars(self) -> pl.Expr:
        return self.lhs.to_polars() > self.rhs.to_polars()


class GeExpression(BinaryOperatorExpression, tag='ge'):
    def to_polars(self) -> pl.Expr:
        return self.lhs.to_polars() >= self.rhs.to_polars()


class EqExpression(BinaryOperatorExpression, tag='eq'):
    def to_polars(self) -> pl.Expr:
        return self.lhs.to_polars() == self.rhs.to_polars()


class LtExpression(BinaryOperatorExpression, tag='lt'):
    def to_polars(self) -> pl.Expr:
        return self.lhs.to_polars() < self.rhs.to_polars()


class LeExpression(BinaryOperatorExpression, tag='le'):
    def to_polars(self) -> pl.Expr:
        return self.lhs.to_polars() <= self.rhs.to_polars()


class NeqExpression(BinaryOperatorExpression, tag='neq'):
    def to_polars(self) -> pl.Expr:
        return self.lhs.to_polars() != self.rhs.to_polars()


# Binary Arithmetic Expressions

class PlusExpression(BinaryOperatorExpression, tag='plus'):
    def to_polars(self) -> pl.Expr:
        return self.lhs.to_polars() + self.rhs.to_polars()


class MinusExpression(BinaryOperatorExpression, tag='minus'):
    def to_polars(self) -> pl.Expr:
        return self.lhs.to_polars() - self.rhs.to_polars()


class MultiplyExpression(BinaryOperatorExpression, tag='multiply'):
    def to_polars(self) -> pl.Expr:
        return self.lhs.to_polars() * self.rhs.to_polars()


class TrueDivExpression(BinaryOperatorExpression, tag='truediv'):
    def to_polars(self) -> pl.Expr:
        return self.lhs.to_polars() / self.rhs.to_polars()


class FloorDivExpression(BinaryOperatorExpression, tag='floordiv'):
    def to_polars(self) -> pl.Expr:
        return self.lhs.to_polars() // self.rhs.to_polars()


# Unary Arithmetic Expressions

class UnaryArithmeticBaseExpression(Expression):
    value: 'AnyExpression'


class Log10Expression(UnaryArithmeticBaseExpression, tag='log10'):
    def to_polars(self) -> pl.Expr:
        return self.value.to_polars().log10()


class LogExpression(UnaryArithmeticBaseExpression, tag='log'):
    def to_polars(self) -> pl.Expr:
        return self.value.to_polars().log()


class Log2Expression(UnaryArithmeticBaseExpression, tag='log2'):
    def to_polars(self) -> pl.Expr:
        return self.value.to_polars().log() / pl.lit(log(2))


class AbsExpression(UnaryArithmeticBaseExpression, tag='abs'):
    def to_polars(self) -> pl.Expr:
        return self.value.to_polars().abs()


class SqrtExpression(UnaryArithmeticBaseExpression, tag='sqrt'):
    def to_polars(self) -> pl.Expr:
        return self.value.to_polars().sqrt()


class FloorExpression(UnaryArithmeticBaseExpression, tag='floor'):
    def to_polars(self) -> pl.Expr:
        return self.value.to_polars().floor()


class RoundExpression(UnaryArithmeticBaseExpression, tag='round'):
    def to_polars(self) -> pl.Expr:
        return self.value.to_polars().round()


class CeilExpression(UnaryArithmeticBaseExpression, tag='ceil'):
    def to_polars(self) -> pl.Expr:
        return self.value.to_polars().ceil()


class UnaryMinusExpression(UnaryArithmeticBaseExpression, tag='negate'):
    def to_polars(self) -> pl.Expr:
        return -self.value.to_polars()  # Unary minus operator

# Type casting


class CastExpression(Expression, tag='cast'):
    value: 'AnyExpression'
    dtype: PType
    strict: typing.Optional[bool] = None

    def to_polars(self) -> pl.Expr:
        return self.value.to_polars().cast(toPolarsType(self.dtype), strict=self.strict or False)

# Boolean Logic Expressions


class AndExpression(Expression, tag='and'):
    operands: list['AnyExpression']

    def to_polars(self) -> pl.Expr:
        polars_operands = [op.to_polars() for op in self.operands]
        if not polars_operands:
            # Define behavior for empty operands: 'and' -> True
            return pl.lit(True)
        return pl.all_horizontal(polars_operands)


class OrExpression(Expression, tag='or'):
    operands: list['AnyExpression']

    def to_polars(self) -> pl.Expr:
        polars_operands = [op.to_polars() for op in self.operands]
        if not polars_operands:
            # Define behavior for empty operands: 'or' -> False
            return pl.lit(False)
        return pl.any_horizontal(polars_operands)


# Not Expression
class NotExpression(Expression, tag='not'):
    value: 'AnyExpression'

    def to_polars(self) -> pl.Expr:
        # Use bitwise NOT operator (~) which acts as logical NOT for boolean expressions in Polars
        return ~self.value.to_polars()


# Null Check Expressions

class IsNaExpression(Expression, tag='is_na'):
    value: 'AnyExpression'

    def to_polars(self) -> pl.Expr:
        return self.value.to_polars().is_null()


class IsNotNaExpression(Expression, tag='is_not_na'):
    value: 'AnyExpression'

    def to_polars(self) -> pl.Expr:
        return self.value.to_polars().is_not_null()


# Column Reference Expression
class ColumnReferenceExpression(Expression, tag='col'):
    name: str

    def to_polars(self) -> pl.Expr:
        return pl.col(self.name)


# Constant Value Expression
class ConstantValueExpression(Expression, tag='const'):
    value: typing.Union[str, int, float, bool, None]

    def to_polars(self) -> pl.Expr:
        return pl.lit(self.value)


# Min/Max Expressions

class MinExpression(Expression, tag='min'):
    operands: list['AnyExpression']

    def to_polars(self) -> pl.Expr:
        polars_operands = [op.to_polars() for op in self.operands]
        if not polars_operands:
            return pl.lit(None)
        return pl.min_horizontal(polars_operands)


class MaxExpression(Expression, tag='max'):
    operands: list['AnyExpression']

    def to_polars(self) -> pl.Expr:
        polars_operands = [op.to_polars() for op in self.operands]
        if not polars_operands:
            return pl.lit(None)
        return pl.max_horizontal(polars_operands)

File: lib/ptabler/schema/src/expressions/basics.ts
import type { AxisSpec } from "@milaboratories/pl-model-common";
import type { DataType } from "../common";
import type { Expression } from "./base";

/** Represents all possible comparison operator types. */
export type ComparisonOperator = "gt" | "ge" | "eq" | "lt" | "le" | "neq";

/** Defines a comparison operation between two expressions. */
export interface ComparisonExpression {
  /** The type of comparison (e.g., 'gt', 'eq'). */
  type: ComparisonOperator;
  /** The left-hand side expression. */
  lhs: Expression;
  /** The right-hand side expression. */
  rhs: Expression;
}

/** Defines the supported binary arithmetic operators. */
export type BinaryArithmeticOperator = "plus" | "minus" | "multiply" | "truediv" | "floordiv";

/** Represents a binary arithmetic operation between two expressions. */
export interface BinaryArithmeticExpression {
  /** The type of arithmetic operation (e.g., 'plus', 'minus'). */
  type: BinaryArithmeticOperator;
  /** The left-hand side expression. */
  lhs: Expression;
  /** The right-hand side expression. */
  rhs: Expression;
}

/** Defines the supported unary arithmetic operators. */
export type UnaryArithmeticOperator =
  | "log10"
  | "log"
  | "log2"
  | "abs"
  | "sqrt"
  | "negate"
  | "floor"
  | "round"
  | "ceil";

/** Represents a unary arithmetic operation on a single expression. */
export interface UnaryArithmeticExpression {
  /** The type of unary operation (e.g., 'log10', 'abs'). */
  type: UnaryArithmeticOperator;
  /** The expression to operate on. */
  value: Expression;
}

/**
 * Represents a type casting operation that converts the result of an expression to a specified data type.
 */
export interface CastExpression {
  /** The type of operation, always 'cast'. */
  type: "cast";
  /** The expression whose result will be cast to the target data type. */
  value: Expression;
  /** The target data type to cast the expression result to. */
  dtype: DataType;
  /**
   * Whether to use strict casting mode. If true, conversion errors and overflows will throw exceptions.
   * If false or undefined, uses non-strict mode where failures result in null values. Defaults to false.
   */
  strict?: boolean;
}

/** Defines the supported boolean list operators. */
export type BooleanListOperator = "and" | "or";

/** Represents a boolean logic operation (AND, OR) on a list of expressions. */
export interface BooleanLogicExpression {
  /** The type of boolean operation ('and', 'or'). */
  type: BooleanListOperator;
  /** An array of boolean expressions as operands. */
  operands: Expression[]; // Array of boolean expressions
}

/** Represents a logical NOT operation on a single boolean expression. */
export interface NotExpression {
  /** The type of operation, always 'not'. */
  type: "not";
  /** The boolean expression to negate. */
  value: Expression;
}

/** Defines the supported null check operators. */
export type NullCheckOperator = "is_na" | "is_not_na";

/** Represents a null check operation (is NA, is not NA) on an expression. */
export interface NullCheckExpression {
  /** The type of null check ('is_na', 'is_not_na'). */
  type: NullCheckOperator;
  /** The expression to check for nullity. */
  value: Expression;
}

/** Represents a reference to a column by its name. */
export interface ColumnReferenceExpression {
  /** The type of operation, always 'col'. */
  type: "col";
  /** The name of the column to reference. */
  name: string;
}

/** Represents a reference to an axis by its specification (or id). */
export interface AxisReferenceExpression {
  /** The type of operation, always 'axis'. */
  type: "axis";
  /** The axis to reference. */
  spec: AxisSpec;
}

/** Represents a constant literal value (string, number, boolean, or null). */
export interface ConstantValueExpression {
  /** The type of operation, always 'const'. */
  type: "const";
  /** The constant value. */
  value: string | number | boolean | null;
}

/** Defines the supported min/max operators. */
export type MinMaxOperator = "min" | "max";

/** Represents a min or max operation on a list of expressions. */
export interface MinMaxExpression {
  /** The type of operation ('min' or 'max'). */
  type: MinMaxOperator;
  /** An array of expressions to find the minimum or maximum value from. */
  operands: Expression[];
}

/**
 * Represents an "in set" operation.
 * Checks if the value of an expression is contained within a specified set of values.
 * Returns true if the expression's value is found in the set, false otherwise.
 */
export interface InSetExpression {
  /** The type of operation, always 'in_set'. */
  type: "in_set";
  /** The expression whose value will be checked for membership in the set. */
  value: Expression;
  /** The set of values to check membership against. */
  set: (string | number | boolean | null)[];
}

/**
 * Represents an alias operation.
 * This operation creates a new expression with a specified alias.
 */
export interface AliasExpression {
  /** The type of operation, always 'alias'. */
  type: "alias";
  /** The expression to alias. */
  value: Expression;
  /** The alias name. */
  name: string;
}
